internal/mesh/internal/controllers/sidecarproxy/builder/local_app.go

```diff
@@ -0,0 +1,141 @@
+package builder
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/consul/envoyextensions/xdscommon"
+	pbcatalog "github.com/hashicorp/consul/proto-public/pbcatalog/v1alpha1"
+	"github.com/hashicorp/consul/proto-public/pbmesh/v1alpha1/pbproxystate"
+)
+
+func (b *Builder) BuildLocalApp(workload *pbcatalog.Workload) *Builder {
+	return b.addInboundListener(xdscommon.PublicListenerName, workload).
+		addInboundRouters(workload).
+		addInboundTLS()
+}
+
+func (b *Builder) getLastBuiltListener() *pbproxystate.Listener {
+	lastBuiltIndex := len(b.proxyStateTemplate.ProxyState.Listeners) - 1
+	return b.proxyStateTemplate.ProxyState.Listeners[lastBuiltIndex]
+}
+
+func (b *Builder) addInboundListener(name string, workload *pbcatalog.Workload) *Builder {
+	listener := &pbproxystate.Listener{
+		Name:      name,
+		Direction: pbproxystate.Direction_DIRECTION_INBOUND,
+	}
+
+	// We will take listener bind port from the workload for now.
+	// Find mesh port.
+	var meshPort string
+	for portName, port := range workload.Ports {
+		if port.Protocol == pbcatalog.Protocol_PROTOCOL_MESH {
+			meshPort = portName
+			break
+		}
+	}
+
+	// Check if the workload has a specific address for the mesh port.
+	var meshAddress string
+	for _, address := range workload.Addresses {
+		for _, port := range address.Ports {
+			if port == meshPort {
+				meshAddress = address.Host
+			}
+		}
+	}
+	// Otherwise, assume the first address in the addresses list.
+	if meshAddress == "" {
+		// It is safe to assume that there's at least one address because we validate it when creating the workload.
+		meshAddress = workload.Addresses[0].Host
+	}
```

Comment:
It might make sense to pull this into a helper. Some parts of this (like defaulting to the first host. Presumably because an empty ports list matches all ports?) don't feel obvious.